File: src/main/GO/leetcode/lists/topinterview/LeetCode150.go
package topinterview

import "strconv"

func evalRPN(tokens []string) int {

	if len(tokens) == 0 {
		return 0
	}

	stack := make([]string, 0)
	for i := 0; i < len(tokens); i++ {
		token := tokens[i]
		one, tow := stack[len(stack)-1], stack[len(stack)-2]
		stack = stack[:len(stack)-2]
		switch token {
		case "+":
			res := Str2Int(one) + Str2Int(tow)
			stack = append(stack, strconv.Itoa(res))
		case "-":
			res := Str2Int(tow) - Str2Int(one)
			stack = append(stack, strconv.Itoa(res))
		case "*":
			res := Str2Int(one) * Str2Int(tow)
			stack = append(stack, strconv.Itoa(res))
		case "/":
			res := Str2Int(tow) / Str2Int(one)
			stack = append(stack, strconv.Itoa(res))
		default:
			stack = append(stack, token)
		}
	}
	return Str2Int(stack[0])
}

File: src/main/java/leetcode/lists/topinterview/LeetCode150.java
package leetcode.lists.topinterview;

import java.util.Stack;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/27 23:55
 */
public class LeetCode150 {

    public int evalRPN(String[] tokens) {

        if (tokens == null || tokens.length == 0) {
            return 0;
        }
        Stack<Integer> stack = new Stack<>();

        for (String s : tokens) {

            switch (s) {
                case "+":
                    stack.push(stack.pop() + stack.pop());
                    break;
                case "-":
                    Integer one = stack.pop();
                    Integer two = stack.pop();
                    stack.push(two - one);
                    break;
                case "*":
                    stack.push(stack.pop() * stack.pop());
                    break;
                case "/":
                    Integer a = stack.pop();
                    Integer b = stack.pop();
                    stack.push(b / a);
                    break;
                default:
                    stack.push(Integer.valueOf(s));
            }
        }
        return stack.pop();
    }
}
